File: tests/e2e/fixtures/medium-ts-project/src/services/data-processor.ts
/**
 * Data processing service
 */

import { ProcessingOptions } from '../config/config-manager';
import { ValidationError, ParsingError, TimeoutError } from '../types/errors';

/**
 * Interface for processed data result
 */
export interface ProcessedData {
  id: string;
  originalData: any;
  processedData: any;
  metadata: {
    processingTime: number;
    timestamp: string;
    steps: string[];
  };
}

/**
 * Service for processing data
 */
export class DataProcessor {
  private activeProcesses: Map<string, Promise<ProcessedData>> = new Map();
  
  /**
   * Creates a new DataProcessor
   * @param options - Processing options
   */
  constructor(private options: ProcessingOptions) {}
  
  /**
   * Processes data according to options
   * @param data - Data to process
   * @returns Processed data
   */
  public async processData(data: any): Promise<ProcessedData> {
    // Generate a process ID
    const processId = this.generateProcessId();
    
    // Create processing promise
    const processingPromise = this.createProcessingPromise(processId, data);
    
    // Store in active processes
    this.activeProcesses.set(processId, processingPromise);
    
    try {
      // Wait for processing to complete
      const result = await processingPromise;
      
      // Remove from active processes
      this.activeProcesses.delete(processId);
      
      return result;
    } catch (error) {
      // Remove from active processes
      this.activeProcesses.delete(processId);
      
      // Re-throw the error
      throw error;
    }
  }
  
  /**
   * Creates a promise that processes the data
   * @param processId - Unique process ID
   * @param data - Data to process
   * @returns Promise that resolves to processed data
   */
  private createProcessingPromise(processId: string, data: any): Promise<ProcessedData> {
    return new Promise<ProcessedData>((resolve, reject) => {
      // Create timeout if configured
      const timeoutId = this.options.timeout > 0
        ? setTimeout(() => {
            reject(new TimeoutError(`Processing timed out after ${this.options.timeout}ms`));
          }, this.options.timeout)
        : null;
      
      // Process the data asynchronously
      this.executeProcessing(processId, data)
        .then(result => {
          // Clear timeout if set
          if (timeoutId) clearTimeout(timeoutId);
          resolve(result);
        })
        .catch(error => {
          // Clear timeout if set
          if (timeoutId) clearTimeout(timeoutId);
          reject(error);
        });
    });
  }
  
  /**
   * Executes the actual data processing
   * @param processId - Unique process ID
   * @param data - Data to process
   * @returns Processed data
   */
  private async executeProcessing(processId: string, data: any): Promise<ProcessedData> {
    const startTime = Date.now();
    const steps: string[] = [];
    
    try {
      // Validate data if enabled
      let validatedData = data;
      if (this.options.validate) {
        steps.push('validation');
        validatedData = await this.validateData(data);
      }
      
      // Transform data if enabled
      let transformedData = validatedData;
      if (this.options.transform) {
        steps.push('transformation');
        transformedData = await this.transformData(validatedData);
      }
      
      // Create the result
      const result: ProcessedData = {
        id: processId,
        originalData: data,
        processedData: transformedData,
        metadata: {
          processingTime: Date.now() - startTime,
          timestamp: new Date().toISOString(),
          steps
        }
      };
      
      return result;
    } catch (error) {
      // Add processing time to error
      const processingTime = Date.now() - startTime;
      
      // Re-throw with additional context
      if (error instanceof Error) {
        error.message = `${error.message} (Process ${processId}, Time: ${processingTime}ms)`;
      }
      
      throw error;
    }
  }
  
  /**
   * Validates input data
   * @param data - Data to validate
   * @returns Validated data
   */
  private async validateData(data: any): Promise<any> {
    // Simulate some async validation
    await new Promise(resolve => setTimeout(resolve, 50));
    
    // Simple validation for demonstration
    if (!data) {
      throw new ValidationError('Data cannot be null or undefined');
    }
    
    if (typeof data === 'string') {
      try {
        // Try to parse as JSON
        return JSON.parse(data);
      } catch (error) {
        throw new ParsingError('Failed to parse string data as JSON', error);
      }
    }
    
    // Return the data if it passes validation
    return data;
  }
  
  /**
   * Transforms the data
   * @param data - Data to transform
   * @returns Transformed data
   */
  private async transformData(data: any): Promise<any> {
    // Simulate async transformation
    await new Promise(resolve => setTimeout(resolve, 100));
    
    // Simple transformation for demonstration
    if (Array.isArray(data)) {
      return data.map(item => ({
        ...item,
        transformed: true,
        timestamp: new Date().toISOString()
      }));
    }
    
    if (typeof data === 'object' && data !== null) {
      return {
        ...data,
        transformed: true,
        timestamp: new Date().toISOString()
      };
    }
    
    // Primitive value, just return as is
    return data;
  }
  
  /**
   * Generates a unique process ID
   * @returns Process ID
   */
  private generateProcessId(): string {
    return `process-${Date.now()}-${Math.random().toString(36).substr(2, 9)}`;
  }
  
  /**
   * Cleans up any resources
   */
  public async cleanup(): Promise<void> {
    // Cancel any active processes
    // In a real app, would have a way to cancel each process
    this.activeProcesses.clear();
  }
}

File: tests/e2e/fixtures/python-project/utils/data_processor.py
"""
Data processing module for the Python test project
"""

import time
import json
import logging
from typing import Any, Dict, List, Union
from datetime import datetime
import threading

from .error_handler import ValidationError, ParsingError, TimeoutError

logger = logging.getLogger(__name__)

def process_data(
    data: Any,
    validate: bool = True,
    transform: bool = True,
    timeout: float = 5.0
) -> Dict[str, Any]:
    """
    Process input data with validation and transformation
    
    Args:
        data: Input data to process
        validate: Whether to validate the data
        transform: Whether to transform the data
        timeout: Timeout in seconds
        
    Returns:
        Processed data with metadata
        
    Raises:
        ValidationError: If validation fails
        ParsingError: If parsing fails
        TimeoutError: If processing times out
    """
    start_time = time.time()
    process_id = f"process-{int(time.time())}-{id(data)}"
    steps = []
    
    logger.info(f"Starting data processing: {process_id}")
    
    # Create a result container that can be accessed from another thread
    result_container = {"result": None, "error": None}
    
    def processing_task():
        try:
            # Validate data if enabled
            validated_data = data
            if validate:
                logger.debug("Validating data")
                steps.append("validation")
                validated_data = _validate_data(data)
            
            # Transform data if enabled
            processed_data = validated_data
            if transform:
                logger.debug("Transforming data")
                steps.append("transformation")
                processed_data = _transform_data(validated_data)
            
            # Create result
            result_container["result"] = {
                "id": process_id,
                "original_data": data,
                "processed_data": processed_data,
                "metadata": {
                    "processing_time": time.time() - start_time,
                    "timestamp": datetime.now().isoformat(),
                    "steps": steps
                }
            }
        except Exception as e:
            logger.error(f"Error in processing task: {e}")
            result_container["error"] = e
    
    # Start processing in a separate thread
    processing_thread = threading.Thread(target=processing_task)
    processing_thread.daemon = True
    processing_thread.start()
    
    # Wait for processing to complete or timeout
    processing_thread.join(timeout)
    
    # Check if processing is still running
    if processing_thread.is_alive():
        # Processing timed out
        error_msg = f"Processing timed out after {timeout} seconds"
        logger.error(error_msg)
        raise TimeoutError(error_msg)
    
    # Check for errors
    if result_container["error"]:
        raise result_container["error"]
    
    # Return the result
    return result_container["result"]

def _validate_data(data: Any) -> Any:
    """
    Validate input data
    
    Args:
        data: Data to validate
        
    Returns:
        Validated data
        
    Raises:
        ValidationError: If validation fails
        ParsingError: If parsing fails
    """
    # Add a small delay to simulate validation
    time.sleep(0.05)
    
    # Check for null data
    if data is None:
        raise ValidationError("Data cannot be None")
    
    # Parse strings as JSON
    if isinstance(data, str):
        try:
            return json.loads(data)
        except json.JSONDecodeError as e:
            raise ParsingError(f"Failed to parse string data as JSON: {e}")
    
    # Return the data if it passes validation
    return data

def _transform_data(data: Any) -> Any:
    """
    Transform the data
    
    Args:
        data: Data to transform
        
    Returns:
        Transformed data
    """
    # Add a small delay to simulate transformation
    time.sleep(0.1)
    
    # Handle lists
    if isinstance(data, list):
        return [{
            **item,
            "transformed": True,
            "timestamp": datetime.now().isoformat()
        } if isinstance(item, dict) else item for item in data]
    
    # Handle dictionaries
    if isinstance(data, dict):
        return {
            **data,
            "transformed": True,
            "timestamp": datetime.now().isoformat()
        }
    
    # Return primitives as is
    return data
